source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -116,6 +129,60 @@ void SimpleHttpCache::insert(const Key& key, Http::ResponseHeaderMapPtr&& respon
   map_[key] = SimpleHttpCache::Entry{std::move(response_headers), std::move(body)};
 }
 
+SimpleHttpCache::Entry
+SimpleHttpCache::varyLookup(const LookupRequest& request,
+                            const Http::ResponseHeaderMapPtr& response_headers) {
+  // This method should be called from lookup, which holds the mutex for reading.
+  mutex_.AssertReaderHeld();
+
+  const Http::HeaderEntry* vary_header = response_headers->get(Http::Headers::get().Vary);
+  ASSERT(vary_header);
+
+  Key varied_request_key = request.key();
+  const std::string vary_key = VaryHeader::createVaryKey(vary_header, request.getVaryHeaders());
+  varied_request_key.add_custom_fields(vary_key);
+
+  auto iter = map_.find(varied_request_key);
+  if (iter == map_.end()) {
+    return SimpleHttpCache::Entry{};
+  }
+  ASSERT(iter->second.response_headers_);
+
+  return SimpleHttpCache::Entry{
+      Http::createHeaderMap<Http::ResponseHeaderMapImpl>(*iter->second.response_headers_),
+      iter->second.body_};
+}
+
+void SimpleHttpCache::varyInsert(const Key& request_key,
+                                 Http::ResponseHeaderMapPtr&& response_headers, std::string&& body,
+                                 const Http::RequestHeaderMapPtr& request_vary_headers) {
+  absl::WriterMutexLock lock(&mutex_);
+
+  const Http::HeaderEntry* vary_header = response_headers->get(Http::Headers::get().Vary);
+  ASSERT(vary_header);
+
+  // Insert the varied response.
+  Key varied_request_key = request_key;
+  const std::string vary_key = VaryHeader::createVaryKey(vary_header, request_vary_headers);
+  varied_request_key.add_custom_fields(vary_key);
+  map_[varied_request_key] = SimpleHttpCache::Entry{std::move(response_headers), std::move(body)};
+
+  // Add a special entry to flag that this request generates varied responses.
+  auto iter = map_.find(request_key);
+  const bool first_entry = (iter == map_.end());
+  if (first_entry) {
+    Http::ResponseHeaderMapPtr vary_only_map =
+        Http::createHeaderMap<Http::ResponseHeaderMapImpl>({});
+    vary_only_map->setCopy(Http::Headers::get().Vary, vary_header->value().getStringView());
+    // TODO(cbdm): In a cache that evicts entries, we could maintain a list of the "varykey"s that
+    // we have inserted as the body for this first lookup. This way, we would know which keys we
+    // have inserted for that resource. For the first entry simply use vary_key as the entry_list,
+    // for future entries append vary_key to existing list.
+    std::string entry_list;
+    map_[request_key] = SimpleHttpCache::Entry{std::move(vary_only_map), std::move(entry_list)};
```

Comment:
It's more efficient to `emplace` things like this.

Reply:
I was just using the same syntax from the `insert` function.
Would you rather have `emplace` for efficiency or like this for uniformity?

Reply:
Consistency's probably best here.